pkgs/top-level/by-name-overlay.nix

```
@@ -0,0 +1,46 @@
+# This file turns the pkgs/by-name directory (see its README.md for more info) into an overlay that adds all the defined packages.
+# No validity checks are done here,
+# instead this file is optimised for performance,
+# and validity checks are done by CI on PRs.
+
+# Type: Path -> Overlay
+baseDirectory:
+let
+  # Because of Nix's import-value cache, importing lib is free
+  lib = import ../../lib;
+
+  inherit (builtins)
+    readDir
+    ;
+
+  inherit (lib.attrsets)
+    mapAttrs
+    mapAttrsToList
+    mergeAttrsList
+    ;
+
+  # Package files for a single shard
+  # Type: String -> String -> AttrsOf Path
+  namesForShard = shard: type:
+    if type != "directory" then
+      # Ignore all non-directories. Technically only README.md is allowed as a file in the base directory, so we could alternatively:
+      # - Assume that README.md is the only file and change the condition to `shard == "README.md"` for a minor performance improvement.
+      #   This would however cause very poor error messages if there's other files.
+      # - Ensure that README.md is the only file, throwing a better error message if that's not the case.
+      #   However this would make for a poor code architecture, because one type of error would have to be duplicated in the validity checks and here.
+      # Additionally in either of those alternatives, we would have to duplicate the hardcoding of "README.md"
+      { }
+    else
+      mapAttrs
+        (name: _: baseDirectory + "/${shard}/${name}/package.nix")
+        (readDir (baseDirectory + "/${shard}"));
+
+  # The attribute set mapping names to the package files defining them
+  # This is defined up here in order to allow reuse of the value (it's kind of expensive to compute)
+  # if the overlay has to be applied multiple times
+  packageFiles = mergeAttrsList (mapAttrsToList namesForShard (readDir baseDirectory));
+in
+self: super:
+mapAttrs (name: file:
+  self.callPackage file { }
+) packageFiles
```

Did you try this? It might release a significant number of thunks, including their closures, including the `readDir` attrsets and such.

```suggestion
) (builtins.deepSeq packageFiles packageFiles)
```

The cost of the computation may or may not pay off thanks/despite reduced gc times later.
As for memory I would expect a reduction, but I haven't tried it. The way to measure it is with the final heap size (`NIX_SHOW_STATS=1`), but unfortunately that measure is rounded and noisy. I think Nix needs to do a full GC before reporting stats...

---

I didn't try this, but I now added a comment suggesting this for a potential future optimisation: [diff](https://github.com/NixOS/nixpkgs/compare/4d49c6b94cc40b4ae1afab8eb33c4f78ef9122cd..ebea45378c8d4bd2efdee343c3422509f64ca7b3#diff-caaafdd48fde2a02cd655ca77794b57e1f72f1edce04b239ab72f212ba9d1608)